api/app/signals/apps/signals/utils/map.py

```diff
@@ -0,0 +1,78 @@
+import io
+import urllib.request
+from math import asinh, pi, radians, tan
+
+from PIL import Image
+
+TILE_SIZE = 256
+
+
+#  https://wiki.openstreetmap.org/wiki/Slippy_map_tilenames#Tile_numbers_to_lon..2Flat._3
+class MapGenerator:
+    def _deg2num(self, lat_deg, lon_deg, zoom):
+        lat_rad = radians(lat_deg)
+        n = 2.0 ** zoom
+        xtile = (lon_deg + 180.0) / 360.0 * n
+        ytile = (1.0 - asinh(tan(lat_rad)) / pi) / 2.0 * n
+        return (xtile, ytile)
+
+    def deg2num(self, lat_deg, lon_deg, zoom):
+        (xtile, ytile) = self._deg2num(lat_deg, lon_deg, zoom)
+        return (int(xtile), int(ytile))
+
+    def deg2num_pixel(self, lat_deg, lon_deg, zoom):
+        (xtile, ytile) = self._deg2num(lat_deg, lon_deg, zoom)
+        xpixel = int((xtile % 1) * TILE_SIZE)
+        ypixel = int((ytile % 1) * TILE_SIZE)
+        return (xpixel, ypixel)
+
+    def load_image(self, url_template, zoom, x, y):
+        img = Image.new("RGBA", (3*TILE_SIZE, 3*TILE_SIZE), 0)
```

Comment:
The magic number 3 is there because the size of the final map image we need (currently `680x250`)? You probably derived this by rounding up 680/256, this is wrong:

I think that horizontally 3 tiles is not enough. If the signal location marker sits near either the left and right boundaries of the central map tile. In that case you will miss about 90 pixels (680 / 2 - 250 = 90). Vertically you can get away with this because a tile is taller than the little map image we need.

Reply:
The 3* tile size is the way to get at least one tile around the selected tile.
Our composed image is therefore 768x768 pixels. We always need to produce an image that is 680pixels wide. **Keep in mind that the marker is not necessarily in the middle of final cropped image**. So if we have 256 pixels in the middle, we always have enough pixels left and right. We need (680 - 256) / 2 =  212 pixels left and right. Since our tiles are 256 pixels, this will always work...

Reply:
We had a brief discussion, for the record: